rewrite-java/src/main/java/org/openrewrite/java/internal/parser/TypeTable.java

```diff
@@ -0,0 +1,412 @@
+ */
+package org.openrewrite.java.internal.parser;
+
+import lombok.RequiredArgsConstructor;
+import lombok.Value;
+import org.jspecify.annotations.Nullable;
+import org.objectweb.asm.*;
+import org.objectweb.asm.util.CheckClassAdapter;
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Incubating;
+import org.openrewrite.java.JavaParserExecutionContextView;
+import org.xerial.snappy.SnappyInputStream;
+import org.xerial.snappy.SnappyOutputStream;
+
+import java.io.*;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.*;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.regex.Pattern;
+
+import static org.objectweb.asm.ClassReader.SKIP_CODE;
+import static org.objectweb.asm.Opcodes.V1_8;
+import static org.openrewrite.java.internal.parser.JavaParserCaller.findCaller;
+
+/**
+ * Type tables are written as a TSV file with the following columns:
+ * <ul>
+ *     <li>groupId</li>
+ *     <li>artifactId</li>
+ *     <li>version</li>
+ *     <li>classAccess</li>
+ *     <li>className</li>
+ *     <li>classSignature</li>
+ *     <li>classSuperclassSignature</li>
+ *     <li>classSuperinterfaceSignatures[]</li>
+ *     <li>access</li>
+ *     <li>memberName</li>
+ *     <li>descriptor</li>
+ *     <li>signature</li>
+ *     <li>parameterNames</li>
+ *     <li>exceptions[]</li>
+ * </ul>
+ * <p>
+ * Descriptor and signature are in <a href="https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.3">JVMS 4.3</a> format.
+ * Because these type tables could get fairly large, the format is optimized for fast record access. TSV was chosen over CSV for its
+ * simplicity and for the reasons cited <a href="https://github.com/eBay/tsv-utils/blob/master/docs/comparing-tsv-and-csv.md">here</a>.
+ * <p>
+ * There is of course a lot of duplication in the class and GAV columns, but compression cuts down on
+ * the disk impact of that and the value is an overall single table representation.
+ */
+@Incubating(since = "8.44.0")
+@Value
+public class TypeTable implements JavaParserClasspathLoader {
+    /**
+     * Verifies that the bytecodes written out for the types represented in a type table
+     * will not be invalid and therefore rejected by the JVM verifier when used in a compilation
+     * step.
+     */
+    public static final String VERIFY_CLASS_WRITING = "org.openrewrite.java.TypeTableClassWritingVerification";
+
+    public static final String DEFAULT_RESOURCE_PATH = "META-INF/rewrite/classpath.tsv.snappy";
+
+    private static final Map<GroupArtifactVersion, Path> classesDirByArtifact = new LinkedHashMap<>();
+
+    public TypeTable(ExecutionContext ctx, Collection<String> artifactNames) {
+        this(ctx, findCaller().getClassLoader().getResourceAsStream(DEFAULT_RESOURCE_PATH), artifactNames);
+    }
+
+    public TypeTable(ExecutionContext ctx, InputStream is, Collection<String> artifactNames) {
+        try (InputStream snappy = new SnappyInputStream(is)) {
```

Comment:
I would recommend that we use LZ4, Zstd, or even just standard DEFLATE. I was hoping we could get rid of Snappy. Not because it is bad (although IMHO LZ4 and Zstd are typically better), but because it is one more dependency with native libs in it.

I understand that we at the moment already have Snappy on the classpath and not LZ4 nor Zstd. But I was working towards getting rid of Snappy in #4641. So how about using DEFLATE from the JDK or adding either of the other two to the classpath?

Reply:
Done in ac31424840560e84d3c461c0cc9fd141d583f548.